dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,758 @@
+    if (!ptr_rings)
+        throw Exception{"Expected a column containing arrays of points when reading rings", ErrorCodes::TYPE_MISMATCH};
+    offset.ring_sizes.assign(ptr_rings->getOffsets());
+    makeDifferences(offset.ring_sizes);
+
+    return ptr_rings->getDataPtr().get();
+}
+
+const IColumn * unrollSimplePolygons(const ColumnPtr & column, Offset & offset)
+{
+    const auto ptr_polygons = typeid_cast<const ColumnArray*>(column.get());
+    if (!ptr_polygons)
+        throw Exception{"Expected a column containing arrays of points", ErrorCodes::TYPE_MISMATCH};
+    offset.polygon_sizes.assign(ptr_polygons->getOffsets());
+    return ptr_polygons->getDataPtr().get();
+}
+
+void getPointsReprByArrays(const IColumn * column, std::vector<IPolygonDictionary::Point> & dest)
+{
+    const auto ptr_points = typeid_cast<const ColumnArray*>(column);
+    const auto ptr_coord = typeid_cast<const ColumnVector<Float64>*>(&ptr_points->getData());
+    if (!ptr_coord)
+        throw Exception{"Expected coordinates to be of type Float64", ErrorCodes::TYPE_MISMATCH};
+    const auto & offsets = ptr_points->getOffsets();
+    for (size_t i = 0; i < offsets.size(); ++i)
+    {
+        if (offsets[i] - (i == 0 ? 0 : offsets[i - 1]) != 2)
+            throw Exception{"All points should be two-dimensional", ErrorCodes::BAD_ARGUMENTS};
+        dest.emplace_back(ptr_coord->getElement(2 * i), ptr_coord->getElement(2 * i + 1));
+    }
+}
+
+void getPointReprByTuples(const IColumn * column, std::vector<IPolygonDictionary::Point> & dest)
+{
+    const auto ptr_points = typeid_cast<const ColumnTuple*>(column);
+    if (!ptr_points)
+        throw Exception{"Expected a column of tuples representing points", ErrorCodes::TYPE_MISMATCH};
+    if (ptr_points->tupleSize() != 2)
+        throw Exception{"Points should be two-dimensional", ErrorCodes::BAD_ARGUMENTS};
+    const auto column_x = typeid_cast<const ColumnVector<Float64>*>(&ptr_points->getColumn(0));
+    const auto column_y = typeid_cast<const ColumnVector<Float64>*>(&ptr_points->getColumn(1));
+    if (!column_x || !column_y)
+        throw Exception{"Expected coordinates to be of type Float64", ErrorCodes::TYPE_MISMATCH};
+    for (size_t i = 0; i < column_x->size(); ++i)
+    {
+        dest.emplace_back(column_x->getElement(i), column_y->getElement(i));
+    }
+}
+
+}
+
+void IPolygonDictionary::extractPolygons(const ColumnPtr &column)
+{
+    Data data = {polygons, ids};
+    Offset offset;
+
+    const IColumn * points_collection = nullptr;
+    switch (input_type)
+    {
+        case InputType::MultiPolygon:
+            points_collection = unrollMultiPolygons(column, offset);
+            break;
+        case InputType::SimplePolygon:
+            points_collection = unrollSimplePolygons(column, offset);
+            break;
+    }
+
+    /** Adding the first empty polygon */
+    data.addPolygon(true);
+
+    std::vector<Point> points;
```

Comment:
Note: we can avoid creating this vector of points. Just iterate them in different way.
But it's ok for now, not necessary to fix.

Reply:
I did this, so that it could be pulled out into a separate function. 
In hindsight, it makes more sense to move addNewPointFrom[...] into those two functions and we won't be creating anything unnecessary, nor will there be any code duplication. 
I'll fix this later then.

Reply:
Fixed, see latest commits.